rfcs/20200505-transactional-fs.md

```diff
@@ -0,0 +1,421 @@
+# Transactional File Systems Support
+
+| Status        | (Proposed / Accepted / Implemented / Obsolete)                                                |
+| :------------ | :-------------------------------------------------------------------------------------------- |
+| **RFC #**     | [NNN](https://github.com/tensorflow/community/pull/NNN) (update when you have community PR #) |
+| **Author(s)** | Sami Kama (kamsami@amazon.com)                                                                |
+| **Sponsor**   | Mihai Maruseac (mihaimaruseac@google.com)                                                     |
+| **Updated**   | 2020-05-05                                                                                    |
+
+## Objective
+
+The aim of this RFC to extend filesystem access support to persistent storage that
+provide transactional access and eventual consistency.
+
+## Motivation
+
+Current persistent storage implementation in Tensorflow relies on certain guarantees that
+existing local file systems provide. But in the age of big data, local filesystems are not always
+sufficient and/or prefetching the data to local system and then uploading after processing can be error-prone
+and harder to implement for end users. Direct access to persistent storage that provides different gurantees is desirable.
+Cloud storage solutions offered by Amazon, Google and others and databases can be examples of such persistent storages.
+Moreover even though local file systems provide certain atomic-like transactions, they are on file level.
+For the use cases like checkpointing, transactions are emulated through creating a temp directory, adding files to there and then
+renaming/moving temp directory to final location. Such operations would also benefit from enhancements proposed in this RFC.
+
+Transactions can also help with some filesystem access inconsistencies that can happen while reading and writing checkpoints. For example in while on thread reading files from a directory other may be modifying the underlying files. This could lead to reading an inconsistent or corrupt set of files by the reader. With transactions, each thread can have different transaction tokens and underlying file system can choose to postpone modification of files by redirecting them to a temporary location and then moving it in place when transactions end.
+
+## User Benefit
```

Comment:
One question I have is how do we expect FSs with transactions to coexist with FSs without transactions.

Do we expect only user-written python code to be made aware of transactions, or do we also expect opkernel C++ code to be transaction-aware? I ask because we do a lot of IO inside opkernels today (summary writing, checkpointing, data reading, etc) and I'd like to understand whether we plan to modify all this to use transactions and how.

Reply:
The C++/C code change will also impact the existing external file systems plugins that are outside of TensorFlow Core repo. The tensorflow/io repo have several: Azfs (Microsoft Azure), Apache Ignite, OSS, and http. The file system plugins also exists for many other places outside of tensorflow org.

Wondering if there is a non-intrusive method to not break implementations for many places. cc @mihaimaruseac

Reply:
Since this proposal is an extension to [Modular File Systems RFC](https://github.com/tensorflow/community/blob/9c248b5603256c0641a05dd202adb2942c8af500/rfcs/20190506-filesystem-plugin-modular-tensorflow.md), it is expected all filesystems use the API, however as I mentioned above, this may be as much as adding extra pointer to their methods and ignoring passed in value. Filesystems should not be forced to act on transactions and allowed to ignore or use a default transaction. 

@yongtang  Since all file systems are expected to migrate plugin api, I don't think an API change is avoidable and transaction tokens can easily be included at that time.

The second point about making C++ code transaction aware is unfortunately more intrusive. I can think of 2 options there

- Transactions are initiated before ops so that ops can call GetTransactionTokenForFile(or equivalent from you proposal). This probably needs to be achieved by adding transaction begin/end ops to the graph, possibly by the python layer that inserts C++ ops. 

- Transaction tokens are made inputs and outputs of these C++ ops, and created by upstream begin transaction ops and finalized by downstream end transaction ops.

@alextp a quick search didn't turn up many ops, are you aware of any other ops doing file io other than save_restore_v2_ops?

Reply:
@samikama ReadFile and WriteFile are some of the C++ kernel ops, mostly in core/ops/io_ops.cc. But tf.data also has quite a few ops that touches file system API implicitly. As those C++ kernel op APIs are mostly frozen, the input args of the ops may not be easily changeable (version compatibility).